src/server/ua_file.c

```diff
@@ -0,0 +1,419 @@
+{
+    UA_UInt32 fileHandle = *(UA_UInt32 *)input[0].data;
+    UA_ByteString *fileContentToWrite = (UA_ByteString*)input[1].data;
+    UA_StatusCode retval = UA_STATUSCODE_GOOD;
+    UA_FileType *fileObject = getFileTypeObject(server, objectId);
+    if(!fileObject)
+    {
+        return UA_STATUSCODE_BADNOTFOUND;
+    }
+    UA_FileInfo *fileInfo = getFileHandleInfo(server, fileObject, fileHandle);
+    if(!fileInfo)
+    {
+        return UA_STATUSCODE_BADINVALIDARGUMENT;
+    }
+    if(((fileInfo->openFileMode != UA_OPENFILEMODE_ERASEEXISTING) ||
+       (fileInfo->openFileMode != (UA_OPENFILEMODE_ERASEEXISTING | UA_OPENFILEMODE_WRITE))) &&
+       (fileInfo->openFileMode != UA_OPENFILEMODE_READ))
+    {
+        for(size_t i = 0; i < fileContentToWrite->length; i++)
+        {
+            UA_file_print(fileInfo->file, "%c", fileContentToWrite->data[i]);
+        }
+        return retval;
+    }
+    retval = UA_STATUSCODE_BADINVALIDSTATE;
+    return retval;
+}
+
+static UA_StatusCode
+UA_Server_setPositionFileCallback(UA_Server *server,
+                                    const UA_NodeId *sessionId, void *sessionHandle,
+                                    const UA_NodeId *methodId, void *methodContext,
+                                    const UA_NodeId *objectId, void *objectContext,
+                                    size_t inputSize, const UA_Variant *input,
+                                    size_t outputSize, UA_Variant *output)
+{
+    return UA_STATUSCODE_BADNOTIMPLEMENTED;
+}
+
+static UA_StatusCode 
+UA_Server_getPositionFileCallback(UA_Server *server,
+                                    const UA_NodeId *sessionId, void *sessionHandle,
+                                    const UA_NodeId *methodId, void *methodContext,
+                                    const UA_NodeId *objectId, void *objectContext,
+                                    size_t inputSize, const UA_Variant *input,
+                                    size_t outputSize, UA_Variant *output)
+{
+    return UA_STATUSCODE_BADNOTIMPLEMENTED;
+}
+
+UA_StatusCode
+setFileMethodCallbacks(UA_Server *server, const UA_NodeId fileObjectNodeId)
+{
+    UA_StatusCode result = UA_STATUSCODE_GOOD;
+    UA_NodeId openNodeId;
+    UA_NodeId readNodeId;
+    UA_NodeId setPositionNodeId;
+    UA_NodeId getPositionNodeId;
+    UA_NodeId writeNodeId;
+    UA_NodeId closeNodeId;
+
+    result |= getFieldNodeId(server, &fileObjectNodeId, &fieldOpenQN, &openNodeId);
+    result |= getFieldNodeId(server, &fileObjectNodeId, &fieldReadQN, &readNodeId);
+    result |= getFieldNodeId(server, &fileObjectNodeId, &fieldSetPositionQN, &setPositionNodeId);
+    result |= getFieldNodeId(server, &fileObjectNodeId, &fieldGetPositionQN, &getPositionNodeId);
+    result |= getFieldNodeId(server, &fileObjectNodeId, &fieldWriteQN, &writeNodeId);
+    result |= getFieldNodeId(server, &fileObjectNodeId, &fieldCloseQN, &closeNodeId);
```

Comment:
This always returns the NodeIds of the Type Object. Shouldn't it be the NodeID of the Object itself? Is it necessary to add explicitly the methods to the Object?